### DISCUSSION TITLE
Encountered undefined: rados.ErrNotConnected error

### DISCUSSION THREAD

**Discussion by hwdef** [ceph/go-ceph#893] [Q&A] [answered]
I create a demo prjoect
```go
package main

import (
	"fmt"

	"github.com/ceph/go-ceph/cephfs/admin"
)

func main() {
	fsAdmin, err := admin.New()
	if err != nil {
		fmt.Println(err)
	}
	subVolumes, err := fsAdmin.ListSubVolumes("volume_name", "group_name")
	if err != nil {
		fmt.Println(err)
	}
	for _, subVolume := range subVolumes {
		fmt.Printf("subVolume: %v", subVolume)
	}
}

```
When i run this program i get the following error
```
[root@2867d29a597a ceph-demo]# go run main.go 
# github.com/ceph/go-ceph/internal/commands
../code/go/pkg/mod/github.com/ceph/go-ceph@v0.22.0/internal/commands/commands.go:12:16: undefined: rados.ErrNotConnected
```
Does anyone know how to fix it

**Comment by anoopcs9** [comment]
Errors are defined from [rados](https://pkg.go.dev/github.com/ceph/go-ceph/rados#ErrNotConnected) package. You can import `"github.com/ceph/go-ceph/rados"` to get rid of the above _undefined_ error.

Even then you may encounter the real `ErrNotConnected` as required rados connection is not made yet.

**Reply by phlogistonjohn** [reply]
Unfortunately, I don't know if that will make a difference. Note that the error is being reported from line 12 of `../code/go/pkg/mod/github.com/ceph/go-ceph@v0.22.0/internal/commands/commands.go` and that's part of go-ceph, not the sample main file. @hwdef can try that out  to see if it works, but I would want to investigate further even if it does.

Personally, I have no idea why/how this might happen. I'd suggest trying to build it again, but in a clean environment like a container because all I can think is that somehow the cached source might be wrong?

**Reply by anoopcs9** [reply]
Ah.. you are correct. @hwdef please ignore my no-look response 😕.

**Comment by phlogistonjohn** [comment]
This is not directly related to your question but I'd like to point out that `admin.New` is deprecated. I strongly suggest not using it in new code.

**Reply by phlogistonjohn** [reply]
Use https://pkg.go.dev/github.com/ceph/go-ceph@v0.22.0/cephfs/admin#NewFromConn  instead

**Comment by hwdef** [comment]
> Unfortunately, I don't know if that will make a difference. Note that the error is being reported from line 12 of `../code/go/pkg/mod/github.com/ceph/go-ceph@v0.22.0/internal/commands/commands.go` and that's part of go-ceph, not the sample main file. @hwdef can try that out to see if it works, but I would want to investigate further even if it does.
> 
> Personally, I have no idea why/how this might happen. I'd suggest trying to build it again, but in a clean environment like a container because all I can think is that somehow the cached source might be wrong?

I created this project in image [1.20rc2-alpine3.17](https://hub.docker.com/layers/library/golang/1.20rc2-alpine3.17/images/sha256-366257e6f6f6ccdf60ba7f1e7e8e6871dbb737f21b890ac4ad01f507bb581e8d?context=explore).

I installed libcephfs, librbd, librados, and then I created this project in /root/ceph-demo with the following information in my go environment variables

```
~/ceph-demo # go env GOPATH
/go
~/ceph-demo # go env GOROOT
/usr/local/go
```

**Comment by hwdef** [comment]
> This is not directly related to your question but I'd like to point out that `admin.New` is deprecated. I strongly suggest not using it in new code.

I modified my code to look like the following, but it looks like my project just doesn't recognize the rados package

![image](https://github.com/ceph/go-ceph/assets/13084946/20c4b2a1-bdcf-4738-8d93-7968b0db8f43)

**Reply by phlogistonjohn** [reply, marked as answer]
In a way, this is good, because it means your original problem and this issue are probably related. Debugging one may end up debugging both.
Somehow your environment is missing or disabling the rados package. Are you doing anything to forcibly disable cgo?  The rados, rbd, and cephfs subpackages all require cgo and only the "admin" packages don't require cgo (and rgw but it's unique anyway). Could that be why your system can import cephfs/admin but not rados?

**Reply by phlogistonjohn** [reply]
Come to think of it, could it be that golang on alpine image automatically disables cgo? I know that alpine has a reputation for being small and that it uses a different libc than glibc by default. Maybe try a different base image.

**Reply by hwdef** [reply]
Yes, you are right. This image turns off cgo by default, when I turn on cgo, it can find rados.
Thank you very much!
